tests/unittest/_torch/visual_gen/multi_gpu/test_ring_attention.py

```
@@ -0,0 +1,524 @@
+        run_test_in_distributed(world_size=4, test_fn=_logic_ring_invalid_mask, use_cuda=True)
+
+
+class TestRingAttentionFlashAttn4:
+    """Ring with FA4 inner (production-style)."""
+
+    def test_ring_fa4_vs_standard(self):
+        run_test_in_distributed(world_size=4, test_fn=_logic_ring_fa4_vs_standard, use_cuda=True)
+
+
+class TestRingUlyssesAttention:
+    """``UlyssesAttention(RingAttention(inner))`` on 4 GPUs (2×2 cp × ulysses)."""
```

Ruff (0.15.12)

[warning] Docstring contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF002)